### PR TITLE
Fix invalid record referencing in records issue

### LINKED ISSUE #12388: Referencing an undefined record in a descriptor results in an NPE
Consider the following program:
```ballerina
public function main() {
}

type Employee record {
    Address adr;
    *Person;
};

type Address record {
    string city;
    string country;
};
```
The above program gives the following compile error:
```
error: .::map.bal:6:6: unknown type 'Person'
````
This also crashes the compiler and the following is logged in ballerina-internal.log
```
[2018-12-04 19:23:25,412] ERROR {org.ballerinalang.launcher.Main} - null 
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.util.diagnotic.BLangDiagnosticLog.reportDiagnostic(BLangDiagnosticLog.java:106)
	at org.wso2.ballerinalang.compiler.util.diagnotic.BLangDiagnosticLog.error(BLangDiagnosticLog.java:76)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolResolver.visit(SymbolResolver.java:1015)
	at org.wso2.ballerinalang.compiler.tree.types.BLangUserDefinedType.accept(BLangUserDefinedType.java:69)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolResolver.resolveTypeNode(SymbolResolver.java:648)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolResolver.resolveTypeNode(SymbolResolver.java:639)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.validateFuncReceiver(SymbolEnter.java:1447)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.visit(SymbolEnter.java:561)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:68)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineNode(SymbolEnter.java:170)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineRecordInitFunction(SymbolEnter.java:1313)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineMembers(SymbolEnter.java:1162)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineConstructs(SymbolEnter.java:241)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.visit(SymbolEnter.java:201)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:159)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineNode(SymbolEnter.java:170)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.definePackage(SymbolEnter.java:163)
	at org.wso2.ballerinalang.compiler.PackageLoader.define(PackageLoader.java:415)
	at org.wso2.ballerinalang.compiler.PackageLoader.loadEntryPackage(PackageLoader.java:309)
	at org.wso2.ballerinalang.compiler.Compiler.lambda$compilePackages$3(Compiler.java:153)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.stream.Streams$StreamBuilderImpl.forEachRemaining(Streams.java:419)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:155)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:179)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:88)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:76)
	at org.ballerinalang.launcher.LauncherUtils.compile(LauncherUtils.java:307)
	at org.ballerinalang.launcher.LauncherUtils.runProgram(LauncherUtils.java:102)
	at org.ballerinalang.launcher.Main$RunCmd.execute(Main.java:278)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.launcher.Main.main(Main.java:66)
```

### PR DESCRIPTION
**Pull request by Shan1024**
## Purpose
Resolves #12388

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -1508,6 +1508,20 @@ private void createDummyTypeDefSymbol(BLangTypeDefinition typeDef, SymbolEnv env
...
+        // Todo - Add more kinds
+        switch (typeDef.typeNode.type.tag) {
+            case TypeTags.RECORD:
+                typeDef.symbol.kind = ((BLangRecordTypeNode) typeDef.typeNode).symbol.kind;
+                ((BLangRecordTypeNode) typeDef.typeNode).symbol.scope = env.scope;
+                break;
+            case TypeTags.OBJECT:
+                typeDef.symbol.kind = ((BLangObjectTypeNode) typeDef.typeNode).symbol.kind;
+                ((BLangObjectTypeNode) typeDef.typeNode).symbol.scope = env.scope;
+                break;
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: We can merge records and objects into one by casting the typeNode to `BStructureTypeNode`

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/varref/record-variable-reference-negative.bal [hunk 2]
```diff
@@ -121,3 +121,12 @@ function testUnknownFields() {
     any unknown2;
     {name, married, age: {age, format, unknown1}, unknown2} = p; // unknown fields
 }
+
+type UserData1 record {
+    *Data;
```
Comment (pubudu91) [inline tests/ballerina-unit-test/src/test/resources/test-src/expressions/varref/record-variable-reference-negative.bal]: Shall we add this test to `ClosedRecordTypeReferenceTest` and `OpenRecordTypeReferenceTest`?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 1 above)
Comment (Shan1024) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Resolved with https://github.com/ballerina-platform/ballerina-lang/pull/12628/commits/0b0d27d10600a5f66c9bcc7fbff4c17a740a5c1b.

File: tests/ballerina-unit-test/src/test/resources/test-src/expressions/varref/record-variable-reference-negative.bal (hunk 2 above)
Comment (Shan1024) [reply tests/ballerina-unit-test/src/test/resources/test-src/expressions/varref/record-variable-reference-negative.bal]: Resolved with https://github.com/ballerina-platform/ballerina-lang/pull/12628/commits/9e563442150aeb314adcb739451841f0ca4f2f49.

**Pull request merged**